lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -99,14 +100,15 @@ export default function SubmitGradeForm({
   const [draftGradeValue, setDraftGradeValue] = useState<string | null>(null);
 
   // Track if current grade has changed compared to what was originally loaded
-  const [hasUnsavedChanges, setHasUnsavedChanges] = useState<boolean>(false);
-  const updateHasUnsavedChanges = useCallback(
-    (hasUnsavedChanges: boolean) => {
-      setHasUnsavedChanges(hasUnsavedChanges);
-      onUnsavedChanges?.(hasUnsavedChanges);
-    },
-    [onUnsavedChanges]
+  const hasUnsavedChanges = useMemo(
+    () => draftGradeValue !== null && draftGradeValue !== grade.data,
+    [draftGradeValue, grade.data]
   );
+  const prevHasUnsavedChanges = useRef(hasUnsavedChanges);
+  if (prevHasUnsavedChanges.current !== hasUnsavedChanges) {
+    onUnsavedChanges?.(hasUnsavedChanges);
```

Comment:
Invoking a callback on render is IMO surprising behavior. The example you link to in the React docs is about updating internal _state_ within the component when props change.

One possibility is to make the `onUnsavedChanges` fire only when the state is changed by the user within the component, like the HTML `input` event, which fires when you type in a text field but not when it is changed by code that assigns to `inputElement.value`. The parent `GradingControls` component can reset its `hasUnsavedChanges` state itself when the selected student changes. Another reason to do this is that if for any reason the `SubmitGradeForm` component got unmounted, we'd still need the state to update properly in the parent.

Reply:
> The parent GradingControls component can reset its hasUnsavedChanges state itself when the selected student changes.

Right. That actually makes a lot of sense.

I'll make sure `onUnsavedChanges` is invoked as a result of actual user interaction, and handle the out-of-sync `hasUnsavedChanges` from the parent.